ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS/DNSActions.hs

```diff
@@ -59,40 +59,55 @@ instance Exception exception => Exception (DNSorIOError exception) where
 -- Resource
 --
 
--- | Evolving resource; We use it to reinitialise the dns library if the
+-- | Evolving resource; We use it to reinitialise the DNS library if the
 -- `/etc/resolv.conf` file was modified.
 --
-newtype Resource m err a = Resource {
-    withResource :: m (Either err a, Resource m err a)
+-- Note: `constantResource` and `retryResource` are written using an approach
+-- inspired by _"The Different Aspects of Monads and Mixins"_, by Bruno C. d S.
+-- Oliveira.
+--
+newtype Resource m a = Resource {
+    withResource :: m (a, Resource m a)
   }
+  deriving Functor
+
+constantResource :: forall m a. Applicative m => a -> Resource m a
+constantResource = fix go
+  where
+    go :: (a -> Resource m a)
+       -> (a -> Resource m a)
+    go this a = Resource $ pure (a, this a)
 
--- | Like 'withResource' but retries until success.
+-- | A 'Resource` which will exhibit the given `Resource` but retry it with
+-- a given delay until a success. On first success it will reset the delays.
 --
-withResource' :: MonadDelay m
-              => Tracer m err
+retryResource :: forall m e a. MonadDelay m
+              => Tracer m e
               -> NonEmpty DiffTime
-              -- ^ delays between each re-try
-              -> Resource m err a
-              -> m (a, Resource m err a)
-withResource' tracer = go
+              -> Resource m (Either e a)
+              -> Resource m a
+retryResource tracer ds0 = fix go ds0
   where
-    dropHead :: NonEmpty a -> NonEmpty a
-    dropHead as@(_ :| [])  = as
-    dropHead (_ :| a : as) = a :| as
-
-    go !delays resource = do
-      er <- withResource resource
-      case er of
-        (Left err, resource') -> do
-          traceWith tracer err
-          threadDelay (NonEmpty.head delays)
-          withResource' tracer (dropHead delays) resource'
-        (Right r, resource') ->
-          pure (r, resource')
-
-constantResource :: Applicative m => a -> Resource m err a
-constantResource a = Resource (pure (Right a, constantResource a))
-
+    -- NOTE: this is more general recursion than what B. Oliveira is using. We
+    -- need 'super' as part of the recursion.  The reason is that we want to
+    -- modify `Resource` itself not a traversal.
+    go :: (NonEmpty DiffTime -> Resource m (Either e a) -> Resource m a)
+       -> (NonEmpty DiffTime -> Resource m (Either e a) -> Resource m a)
+    go this ds@(d :| _) super = Resource $ do
```

Comment:
The way you arranged the parentheses could see this as
essentially of type `Mixin (NonEmpty DiffTime -> Resource
m (Either e a) -> Resource m a)`, at least according to
the paper. Shouldn't then this be `go super ds@(d :| _) this`  ?

Reply:
```hs
type Mixin a = a -> a -> a

mix :: Mixin a -> Mixin a -> Mixin a
mix f g = \super this -> f (g super this) this

retryMixin :: forall m e a. Monad m
           => Tracer m e
           -> NonEmpty DiffTime
           -> Mixin (NonEmpty DiffTime -> Resource m (Either e a) -> Resource m a)
retryMixin tracer ds0 = \_super this ds super -> Resource $ do
    ea <- withResource super
    case ea of
      (Left e, super') -> do
        traceWith tracer e
        -- threadDelay
        withResource (this ds super')
      (Right a, super') -> do
        return (a, this ds0 super')


```

the thing is `_super` is useless here, and I don't know if there's a way to implement through inheritance (using `_super` rather than `super` above).

Reply:
> Shouldn't then this be go super ds@(d :| _) this ?

I think I should call `super` something else, since it plays a different role than `super` in the paper.

Reply:
What itches me about this is that it seems to me one is just trying to make the
type tetris work and fit this particular mixin type signatures, when in truth
what the function is actually doing is not related to mixins at all. This
function will be retrying a monadic until it succeeds, it there isn't any
inheritance going on. It `fix` works great here because it does exactly what we
want.

I think this rewrite and the mention of mixins paper is a little bit odd and could be confusing for someone reading this for the first time

Reply:
Indeed, that's why I avoided the `Mixin` in the PR, we don't need inheritance in our cases.  I posted the code only to show that this fits into the `Mixin` pattern (in a trivial way).